### PR TITLE
Labels and labeled old expressions

### PR DESCRIPTION
**Pull request by Felalolf**
Adds labels and labeled old expressions.
In Go, lables have a different namespace than variables. Therefore, I added an additional name resolution.
For labeled old expressions to work, I also had to change how variables are initialized. We now have an explicit initialization statement. Furthermor, I changed how blocks are handled in the desugarer.

### CODE REVIEW

File: src/main/scala/viper/gobra/frontend/Desugar.scala [hunk 1]
```diff
@@ -734,11 +729,33 @@ object Desugar {
...
       val src: Meta = meta(stmt)
 
-      // Generates a fresh variable if `idn` is a wildcard or returns the variable
-      // associated with idn otherwise
-      def getVar(idn: PIdnNode)(t: in.Type): in.AssignableVar = idn match {
-        case _: PWildcard => freshExclusiveVar(t.withAddressability(Addressability.Exclusive))(src)
-        case x => assignableVarD(ctx)(x)
+      /**
+        * Desugars the left side of an assignment.
+        * If the argument is a variable use, Left is returned, containing the desugared variable.
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/frontend/Desugar.scala:734]: This function does not return `Either`

File: src/main/scala/viper/gobra/frontend/Parser.scala [hunk 2]
```diff
@@ -459,11 +459,13 @@ object Parser {
...
         block |
         simpleStmt |
+        labeledStmt |
+        expressionStmt |
         emptyStmt
 
 
     lazy val simpleStmt: Parser[PSimpleStmt] =
-      sendStmt | assignmentWithOp | assignment | shortVarDecl | expressionStmt
+      sendStmt | assignmentWithOp | assignment | shortVarDecl // expressionStmt is parsed separately
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/frontend/Parser.scala:468]: was this change necessary due to trying to parse `labelDef` as an expression?

File: src/main/scala/viper/gobra/translator/Names.scala [hunk 3]
```diff
@@ -10,6 +10,7 @@ import viper.silver.{ast => vpr}
 
 object Names {
   def returnLabel: String = "returnLabel"
+  def labeledVar(base: String, label: String): String = s"${base}_${label}L"
```
Comment (ArquintL) [inline src/main/scala/viper/gobra/translator/Names.scala:13]: what is a labeled variable? resp. for what is it used?

**Review by ArquintL — APPROVED** [review]
Looks good, I only have a few smaller remarks

File: src/main/scala/viper/gobra/frontend/Desugar.scala [hunk 4]
```diff
@@ -734,11 +729,33 @@ object Desugar {
...
+              _ <- declare(v)
+              _ <- write(in.Initialization(v)(src))
+            } yield v
+
+          case unk: PIdnUnk if info.isDef(unk) =>
+            val v = assignableVarD(ctx)(idn).asInstanceOf[in.LocalVar]
+            for {
+              _ <- declare(v)
+              _ <- write(in.Initialization(v)(src))
+            } yield v
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/Desugar.scala:752]: These cases have the same body, maybe you can merge them

File: src/main/scala/viper/gobra/frontend/info/base/SymbolTable.scala [hunk 5]
```diff
@@ -245,4 +238,12 @@ object SymbolTable extends Environments[Entity] {
...
   case class BuiltInMPredicate(tag: BuiltInMPredicateTag, rep: PNode, context: ExternalTypeInfo) extends BuiltInGhostEntity with BuiltInMethodLike with GhostTypeMember
 
+
+
+
+  /**
+    * Label
+    */
+
+  case class Label(decl: PLabeledStmt, ghost: Boolean) extends Entity with Product
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/info/base/SymbolTable.scala:248]: What's the purpose of having a `ghost` flag here? Shouldn't a Label be always non-ghost, even if the labeled statement is ghost?

Comment (jcp19) [reply src/main/scala/viper/gobra/frontend/info/base/SymbolTable.scala:248]: I think I see why now. Is it because a label may occur inside a ghost function or block, in which case you want to ensure that regular control flow does not jump to a ghost statement?

**Review by jcp19 — APPROVED** [review]

**Comment by jcp19** [comment]
Fixes #37 and #39

**Comment by jcp19** [comment]
For some reason, I cannot link issues #37 and #39 so that they are closed automatically when we merge this PR.

File: src/main/scala/viper/gobra/frontend/Parser.scala (hunk 2 above)
Comment (Felalolf) [reply src/main/scala/viper/gobra/frontend/Parser.scala:468]: yes

File: src/main/scala/viper/gobra/frontend/info/base/SymbolTable.scala (hunk 5 above)
Comment (Felalolf) [reply src/main/scala/viper/gobra/frontend/info/base/SymbolTable.scala:248]: yes

**Comment by jcp19** [comment]
This is fine by me, you can merge it
